src/core.ts

```diff
@@ -0,0 +1,339 @@
+  }
+
+  const setExecutorPort = (port: string) => {
+    localStorage.setItem("ad4minPort", port);
+  }
+
+  const setExecutorToken = (token: string) => {
+    localStorage.setItem("ad4minToken", token);
+  }
+
+  const client = new Client({
+    ...args,
+    port,
+    token,
+    url,
+    setExecutorUrl: args.setExecutorUrl || setExecutorUrl,
+    setExecutorPort: args.setExecutorPort || setExecutorPort,
+    setExecutorToken: args.setExecutorToken || setExecutorToken
+  });
+
+  return client;
+}
+
+type Ad4mConnectOptions = {
+  appName: string,
+  appDesc: string,
+  appDomain: string,
+  appIconPath?: string,
+  capabilities: {[x: string]: any}[],
+  dataPath?: string,
+  port?: number,
+  token?: string,
+  url?: string,
+  setExecutorUrl?: (val: string) => void,
+  setExecutorPort?: (val: string) => void,
+  setExecutorToken?: (val: string) => void
+}
+
+type PortSearchStateType = "na" | "searching" | "found" | "not_found";
+type ClientEvents = 'port_found' | 'port_searching' | 'port_notfound' | 'connected_without_capabilities' 
+  | 'connected_with_capabilities' | 'disconnected' | 'agent_locked' | 'capabilties_not_matched' 
+  | 'not_connected' | 'loading' | string;
+
+class Client {
+  apolloClient?: ApolloClient<NormalizedCacheObject>;
+  ad4mClient?: Ad4mClient;
+  requestId?: string;
+  isFullyInitialized = false;
+  port = 12000;
+  portSearchState: PortSearchStateType = "na";
+  appName: string;
+  appDesc: string;
+  appDomain: string;
+  url: string;
+  capabilities: {[x: string]: any}[];
+  listeners: {[x: ClientEvents]: [(...args: any[]) => void]} = {};
+  setExecutorUrl: (val: string) => void;
+  setExecutorPort: (val: string) => void;
+  setExecutorToken: (val: string) => void;
+
+  // @fayeed - params
+  constructor({ appName, appDesc, appDomain, capabilities, port, token, url, setExecutorUrl, setExecutorPort, setExecutorToken }: Ad4mConnectOptions) {
+    //! @fayeed - make it support node.js
+    this.appName = appName;
+    this.appDesc = appDesc;
+    this.appDomain = appDomain;
+    this.capabilities = capabilities;
+    this.setExecutorUrl = setExecutorUrl;
+    this.setExecutorPort = setExecutorPort;
+    this.setExecutorToken = setExecutorToken;
+    if (port) this.port = port!;
+    if (url) this.url = url;
+    if (token && token.length > 0){
+      this.setToken(token!);
+    }
+
+    if (localStorage.getItem("ad4minURL") && token.length > 0) {
+      setTimeout(() => {
+        this.callListener('loading');
+      }, 0)
+      this.buildClient();
+      this.checkConnection();
+    } else {
+      setTimeout(() => {
+        this.callListener('init');
+      }, 500)
+    }
+  }
+
+  async connectRemote(url: string) {
+    this.url = url;
+    this.setExecutorUrl(url);
+
+    this.callListener('loading');
+
+    this.buildClient();
+
+    await this.checkConnection();
+  }
+
+  async connectToPort() {
+    try {
+      this.callListener('loading');
+      const port = 12000; //await checkPort(this.port);
```

Comment:
should this be calling the checkPort function?

Reply:
ahh yeah